rmw_fastrtps_cpp/src/listener_thread.cpp

```diff
@@ -0,0 +1,170 @@
+    }
+  }
+  return RMW_RET_ERROR;
+}
+
+rmw_ret_t
+rmw_fastrtps_cpp::join_listener_thread(rmw_context_t * context)
+{
+  auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+  common_context->thread_is_running.exchange(false);
+  rmw_ret_t rmw_ret = rmw_trigger_guard_condition(common_context->listener_thread_gc);
+  if (RMW_RET_OK != rmw_ret) {
+    return rmw_ret;
+  }
+  try {
+    common_context->listener_thread.join();
+  } catch (const std::exception & exc) {
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to join std::thread: %s", exc.what());
+    return RMW_RET_ERROR;
+  } catch (...) {
+    RMW_SET_ERROR_MSG("Failed to join std::thread");
+    return RMW_RET_ERROR;
+  }
+  rmw_ret = rmw_destroy_guard_condition(common_context->listener_thread_gc);
+  if (RMW_RET_OK != rmw_ret) {
+    return rmw_ret;
+  }
+  return RMW_RET_OK;
+}
+
+#define TERMINATE(msg) \
+  do { \
+    fprintf( \
+      stderr, \
+      RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__function__) ":" \
+      RCUTILS_STRINGIFY(__LINE__) RCUTILS_STRINGIFY(msg) ": %s, terminating ...", \
+      rmw_get_error_string().str); \
+    std::terminate(); \
+  } while (0)
+
+void
+node_listener(rmw_context_t * context)
+{
+  assert(nullptr != context);
+  assert(nullptr != context->impl);
+  assert(nullptr != context->impl->common);
+  auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+  while (common_context->thread_is_running.load()) {
+    assert(nullptr != common_context->sub);
+    assert(nullptr != common_context->sub->data);
+    void * subscriptions_buffer[] = {common_context->sub->data};
+    void * guard_conditions_buffer[] = {common_context->listener_thread_gc->data};
+    rmw_subscriptions_t subscriptions;
+    rmw_guard_conditions_t guard_conditions;
+    subscriptions.subscriber_count = 1;
+    subscriptions.subscribers = subscriptions_buffer;
+    guard_conditions.guard_condition_count = 1;
+    guard_conditions.guard_conditions = guard_conditions_buffer;
+    // number of conditions of a subscription is 2
+    rmw_wait_set_t * wait_set = rmw_create_wait_set(context, 2);
+    if (nullptr == wait_set) {
+      TERMINATE("failed to create wait set");
+    }
+    if (RMW_RET_OK != rmw_wait(
+        &subscriptions,
+        &guard_conditions,
+        nullptr,
+        nullptr,
+        nullptr,
+        wait_set,
+        nullptr))
+    {
+      TERMINATE("rmw_wait failed");
+    }
+    if (subscriptions_buffer[0]) {
+      rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+      bool taken;
+      if (RMW_RET_OK != rmw_take(
+          common_context->sub,
+          static_cast<void *>(&msg),
+          &taken,
+          nullptr))
+      {
+        TERMINATE("rmw_take failed");
+      }
+      if (taken) {
+        if (std::memcmp(
+            reinterpret_cast<char *>(common_context->gid.data),
+            reinterpret_cast<char *>(&msg.gid.data),
+            RMW_GID_STORAGE_SIZE) == 0)
```

Comment:
Maybe there should be a function for comparing the gid, rather than just memcmp being used everywhere.